addons/website_sale/controllers/main.py

```diff
@@ -1225,3 +1225,56 @@ def products_autocomplete(self, term, options={}, **kwargs):
                 res_product['price'] = FieldMonetary.value_to_html(res_product['price'], monetary_options)
 
         return res
+
+    # --------------------------------------------------------------------------
+    # Products Recently Viewed
+    # --------------------------------------------------------------------------
+    @http.route('/shop/products/recently_viewed', type='json', auth='public', website=True)
+    def products_recently_viewed(self, **kwargs):
+        """
+        Returns list of recently viewed products according to current user and product options
+        """
+        excluded_ids = []
+        current_sale_order = request.session.get('sale_order_id')
+        if current_sale_order:
+            order_lines = request.env['sale.order.line'].sudo().search([('order_id', '=', current_sale_order)])
+            excluded_ids += order_lines.mapped('product_id.id')
+        if not request.env.user._is_public():
+            partner_id = request.env.user.partner_id
+            visitor_sudo = request.env['website.visitor'].sudo().with_context(active_test=False).search([('user_partner_id', '=', partner_id.id)])
+        else:
+            visitor_sudo = request.env['website.visitor']._get_visitor_from_request()
+        if visitor_sudo:
+            length = 12 + len(excluded_ids)
+            products = visitor_sudo.website_track_ids.product_id[:length]
+            if excluded_ids:
+                products = products.filtered(lambda r: r.id not in excluded_ids)[:12]
+            if products:
+                fields = ['id', 'name', 'website_url']
+                res = {
+                    'products': products.read(fields),
+                }
+
+                FieldMonetary = request.env['ir.qweb.field.monetary']
+                monetary_options = {
+                    'display_currency': request.website.get_current_pricelist().currency_id,
+                }
+                rating = request.website.viewref('website_sale.product_comment').active
+                for res_product, product in zip(res['products'], products):
+                    combination_info = product._get_combination_info_variant()
+                    res_product.update(combination_info)
+                    res_product['list_price'] = FieldMonetary.value_to_html(res_product['list_price'], monetary_options)
+                    res_product['price'] = FieldMonetary.value_to_html(res_product['price'], monetary_options)
```

Comment:
```
combination_info = product._get_combination_info_variant()
combination_info.update({
    'list_price': FieldMonetary.value_to_html(res_product['list_price'], monetary_options),
    'price': FieldMonetary.value_to_html(res_product['price'], monetary_options)
})
res_product.update(combination_info)
```
or , if list price and list are not linked to combinations
```
combination_info = product._get_combination_info_variant()
res_product.update(combination_info)
res_product.update({
    'list_price': FieldMonetary.value_to_html(res_product['list_price'], monetary_options),
    'price': FieldMonetary.value_to_html(res_product['price'], monetary_options)
})
```

Reply:
not sure to see your change... If it is simply to make code beautiful,

That make no sense to recreate a new dict for it. So no one better than other

Reply:
and don't forget to add a `,` at the last line... in this case !